src/httpd_jsonapi.c

```diff
@@ -1812,16 +1825,117 @@ jsonapi_reply_queue_tracks_add(struct httpd_request *hreq)
       if (pos >= 0)
 	pos += count;
 
-      ttl_count += count;
+      *total_count += count;
     }
-  while ((uri = strtok(NULL, ",")));
+  while ((uri = strtok_r(NULL, ",", &ptr)));
 
   free(uris);
 
+  return ret;
+}
+
+static int
+queue_tracks_add_byexpression(const char *param, int pos, int *total_count)
+{
+  char *expression;
+  struct smartpl smartpl_expression;
+  struct query_params query_params;
+  struct player_status status;
+  int ret;
+
+  memset(&query_params, 0, sizeof(struct query_params));
+
+  query_params.type = Q_ITEMS;
+  query_params.sort = S_ALBUM;
+  query_params.idx_type = I_NONE;
+
+  memset(&smartpl_expression, 0, sizeof(struct smartpl));
+  expression = safe_asprintf("\"query\" { %s }", param);
+  ret = smartpl_query_parse_string(&smartpl_expression, expression);
+  free(expression);
+
+  if (ret < 0)
+    return -1;
+
+  query_params.filter = strdup(smartpl_expression.query_where);
+  free_smartpl(&smartpl_expression, 1);
+
+  player_get_status(&status);
+
+  ret = db_queue_add_by_query(&query_params, status.shuffle, status.item_id, pos, total_count, NULL);
+
+  free(query_params.filter);
+
+  return ret;
+
+}
+
+static int
+jsonapi_reply_queue_tracks_add(struct httpd_request *hreq)
+{
+  const char *param_pos;
+  const char *param_uris;
+  const char *param_expression;
+  const char *param;
+  int pos = -1;
+  bool shuffle;
+  int total_count = 0;
+  json_object *reply;
+  int ret = 0;
+
+
+  param_pos = evhttp_find_header(hreq->query, "position");
+  if (param_pos)
+    {
+      if (safe_atoi32(param_pos, &pos) < 0)
+        {
+	  DPRINTF(E_LOG, L_WEB, "Invalid position parameter '%s'\n", param_pos);
+
+	  return HTTP_BADREQUEST;
+	}
+
+      DPRINTF(E_DBG, L_WEB, "Add tracks starting at position '%d\n", pos);
+    }
+
+  param_uris = evhttp_find_header(hreq->query, "uris");
+  param_expression = evhttp_find_header(hreq->query, "expression");
+
+  if (!param_uris && !param_expression)
+    {
+      DPRINTF(E_LOG, L_WEB, "Missing query parameter 'uris' or 'expression'\n");
+
+      return HTTP_BADREQUEST;
+    }
+
+  // if query parameter "clear" is "true", stop playback and clear the queue before adding new queue items
+  param = evhttp_find_header(hreq->query, "clear");
+  if (param && strcmp(param, "true") == 0)
+    {
+      player_playback_stop();
```

Comment:
Maybe have this in its own function, which can also be called by jsonapi_reply_queue_clear()? It would be to avoid the risk that some future coder makes a modification in jsonapi_reply_queue_clear() and forgets to make the same modification here.